### PR TITLE
feat: activation intrinsics for neural networks

### PR DESCRIPTION
**Pull request by jalvesz**
Add [activation functions](https://en.wikipedia.org/wiki/Activation_function) for neural networks. Relates to https://github.com/fortran-lang/stdlib/issues/858

Todo:

- [x] specs & autodoc
- [x] tests
- [x] examples

**Previous Art**
SciPy Special functions collection: https://docs.scipy.org/doc/scipy/reference/special.html
torch nn: https://pytorch.org/docs/stable/nn.html#non-linear-activations-other
neural-fortran : https://github.com/modern-fortran/neural-fortran/blob/main/src/nf/nf_activation.f90

cc: @Beliavsky @milancurcic @epagone @jvdp1 @perazz

### CODE REVIEW

**Comment by jalvesz** [comment]
Before moving forward, any opinions on 
* putting these functions within the "specialfunctions" category?
* naming of the derivative/gradient version with `<name>_grad`?
* any other remark?

File: src/stdlib_specialfunctions.fypp [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+            ${rt}$ :: y
+        end function
+        #:endfor
+    end interface
+    public :: tanh
+
+    interface tanh_grad
+        !! Version: experimental
+        !!
+        !! gradient of the gaussian function
```
Comment (perazz) [inline src/stdlib_specialfunctions.fypp:334]:
```suggestion
        !! gradient of the hyperbolic tangent function
```

File: src/stdlib_specialfunctions.fypp [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+        #:endfor
+    end interface
+    public :: Softplus_grad
+
+    interface ftanh 
+        !! Version: experimental
+        !!
+        !! Fast approximation of the tanh function
+        !! Source: https://fortran-lang.discourse.group/t/fastgpt-faster-than-pytorch-in-300-lines-of-fortran/5385/31
+        #:for rk, rt in REAL_KINDS_TYPES
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/stdlib_specialfunctions.fypp:443]: Also: I believe a discussion would be needed for fast, approximate versions of intrinsic functions. Different from `gamma` (that returns similar accuracy as the intrinsics), these are fast approximations. I'm fine with the `f` prefix (or `fast_` would also be OK). I believe that should be standardized somewhere, for example in the library style guide?

File: src/stdlib_specialfunctions.fypp [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        #:for rank in RANKS
+        pure module function Softmax_grad_r${rank}$_${rk}$( x , dim ) result( y )
+            ${rt}$, intent(in) :: x${ranksuffix(rank)}$
+            ${rt}$ :: y${shape('x', rank)}$
+            integer, intent(in), optional :: dim
+        end function
+        #:endfor
+        #:endfor
+    end interface
+    public :: Softmax_grad
```
[the remark this answers is not in the document]
Comment (jalvesz) [reply src/stdlib_specialfunctions.fypp]: done here https://github.com/fortran-lang/stdlib/pull/860/commits/66e662ccc6b8d7f0cf90e930f420fe3fb270fcfb

File: include/common.fypp [hunk 4]
```diff
@@ -230,6 +230,29 @@ ${prefix + joinstr.join([line.strip() for line in txt.split("\n")]) + suffix}$
...
+#!
+#! Args:
+#!   varname (str): Name of the variable to be used as origin
+#!   origrank (int): Rank of the original variable
+#!
+#! Returns:
+#!   Shape expression enclosed in braces, so that it can be used as suffix to
+#!   define array shapes in declarations.
+#!
+#:def shape(varname, origrank)
```
[the remark this answers is not in the document]
Comment (jalvesz) [reply include/common.fypp]: replaced by this macro here https://github.com/fortran-lang/stdlib/pull/860/commits/ea74c87fb4dcfc4a1ad8394f73bd1a96d938eda0

**Comment by jalvesz** [comment]
Thanks @perazz for your review! regarding tanh and erf, I wonder if I should actually remove reference to the intrinsic names and simply leave a reference to the fast approximation as that is what makes sense for NNs. Also, should these functions stay here or moved to the `intrinsics` module?

Regarding
>I'm fine with the f prefix (or fast_ would also be OK). I believe that should be standardized somewhere, for example in the library style guide?

I agree with you and also wonder which one would be preferable, I don't have strong oppinions on that.

**Comment by jalvesz** [comment]
@perazz so for the fast functions I propose here to use the naming `fast_<>` as I felt that a `stdlib_<>` feels like stdlib's reference implementation, but here we are proposing degraded accuracy for the sake of activation functions. I've added the documentation for those and also extended the tests for different accuracies but leaving the tolerance fixed to a low accuracy value.

**Comment by Beliavsky** [comment]
A rigorous test would be to print the activation function and derivative values for a range of arguments, write them to a file, and compare them with a reference implementation in Python. How feasible is that?

I have looked at the code and checked a few of the activation functions and derivatives. They look correct. There are some functions with an optional dim argument that is not referenced, for example

```fortran
pure module function softmax_r1_sp( x , dim ) result( y )
    real(sp), intent(in) :: x(:)
    real(sp) :: y(size(x))
    integer, intent(in), optional :: dim

    y = exp(x - maxval(x))
    y = y / sum(y)
end function
```
Can unused dim arguments be removed from functions without too much manual work?

**Comment by jalvesz** [comment]
I'll merge this one that has been hanging for several months. I'll add an issue to enhance the tests with an .npy data-base later on.
